### ISSUE TITLE
Serializing a dictionary in iOS hangs

### ISSUE THREAD

**Issue by rachelkang** [dotnet/macios#13628]
From DevCom issue https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1453815

This is the initial report from the user:

> 
> In iOS with a release build (it works fine with debug build) if you try to serialize a dictionary with Newtonsoft the app hangs.
> 
> I’ve reproduced this easily by adding the following code to my AppDelegate’s FinishedLaunching method:
> 
> Console.WriteLine($"######## A");
> 	var dict = new System.Collections.Generic.Dictionary<string, DateTime>();
> 	var json = Newtonsoft.Json.JsonConvert.SerializeObject(dict);
> 	Console.WriteLine($"######## {json}");
> 
> If you run with this you will see the first console message logged, but the second will never happen and the app will crach after 17 seconds (because it took too long to launch).
> 
> After a bit more investigation I’ve found that enabling the Use the LLVM optimizing compiler option causes this problem. If I disable that then it runs fine. I still don’t know the root cause, and would like to enable LLVM for my release builds.
> 
> I’m continuing to investigate and have found that in addition to LLVM being a factor, having a Notification Content Extension project in my app is also a factor. That is, if I remove that extension then the problem goes away. Interestingly, if I create a new, default Xamarin app then this is not the case. Additionally, if I create a new, default Notification Content Extension project then the problem still occurs. To summarize:
> My app: LLVM compilation with Notification Content Extension causes the problem.
> Default app: LLVM compilation with Notification Content Extension causes no problem.
>

**Comment by rachelkang** [comment]
Hi Rainer, I tried reproducing the issue you've described, but did not see any crash.

Please let me know if this repro sample works for you too: [SerializingDictionaryIssue.zip](https://github.com/xamarin/xamarin-macios/files/7764093/SerializingDictionaryIssue.2.zip)

If so, please provide a sample project and/or full build logs so that we can investigate further.

**Comment by Qythyx** [comment]
@rachelkang, thanks for looking at this. I also tried creating a simple repro with a new project and it did not have the problem. I've been working on taking my app and removing various pieces to make it simple enough so I can share it, and to help identify what part is causing the problem. I will reply here when I have a repro I can share.

**Comment by Qythyx** [comment]
Ok, I simplified my app to a small repro (attached). If you look at `AppDelegate` you will see some very simple code. When I run the app I see debug message A and B, but then the app freezes and the 3rd debug message is never shown. This only happens for a release build (not debug) and only if LLVM is enabled.

[SerializationBug.zip](https://github.com/xamarin/xamarin-macios/files/7767084/SerializationBug.zip)

**Comment by rachelkang** [comment]
Hi, @Qythyx - we tried running your new sample on an iPhone 13 device with LLVM enabled and was still unable to reproduce it. What devices/models and versions of iOS are you experiencing this with?

**Comment by Qythyx** [comment]
@rachelkang , thanks for trying to reproduce this. I'm using an iPhone 12 Pro Max running iOS 15.2. Just to confirm, when you tried to repro you were doing it with a Release build, right?

Is there anything I can do on my side to help you investigate? Are there some logs I could provide or anything like that?

**Comment by rachelkang** [comment]
Yes, I've been testing in Release mode. And thanks for letting us know - we'll try to reproduce it on the same device as soon as we can locate one!

And yes, if you share your [full build logs](https://docs.microsoft.com/visualstudio/msbuild/obtaining-build-logs-with-msbuild?view=vs-2022#save-a-binary-log), it could be helpful. We can take a look at how yours might vary from ours.

**Comment by baskren** [comment]
I'm seeing this as well.  Any updates?

**Comment by rolfbjarne** [comment]
Try adding `--assembly-build-target=@all=staticobject` to the additional mtouch arguments in both the main project and the executable project.

This is the same as adding the following in the csproj for the Release|iPhone section:

```xml
<MtouchExtraArgs>--assembly-build-target=@all=staticobject</MtouchExtraArgs>
```
